kong/pdk/response.lua

```diff
@@ -520,7 +617,12 @@ local function new(self, major_version)
   -- as-is.  It is the caller's responsibility to set the appropriate
   -- Content-Type header via the third argument.  As a convenience, `body` can
   -- be specified as a table; in which case, it will be JSON-encoded and the
-  -- `application/json` Content-Type header will be set.
+  -- `application/json` Content-Type header will be set. On gRPC we cannot send
+  -- the `body` with this function at the moment at least, so what it does
+  -- instead is that it sends "body" in `grpc-message` header instead. If the
+  -- body is a table it looks for a field `message` in it, and uses that as a
+  -- `grpc-message` header. Though, if you have specified `Content-Type` header
+  -- with exact or prefix of `application/grpc`` the body will be send.
```

Comment:
typo/grammar:

```suggestion
  -- starting with `application/grpc`, the body will be sent.
```

However, I think this behavior is a little odd. What happens if you explicitly give it an explicit `Content-Type` _and_ give it a table? In my reading of the code, it looks like it will try to `ngx.print` the value of `body.message` — this looks like an unintended side-effect.

I think `body` should only be sent as a gRPC payload if it is given as a string, and regardless of the explicitly-given `Content-Type`.

Reply:
Fixed the grammar issue. On the gRPC response, the intent was to handle the places (mostly error reporting) where Kong responds with a `{"message": "some message"}`; in these cases, the message must be sent as `grpc-message`, plus a `grpc-status` (obtained [from the HTTP response code](https://github.com/grpc/grpc/blob/master/doc/statuscodes.md) -- at least until richer error reporting (with protobuf body) is supported. I do agree that this method can be further improved for gRPC -- plugins testing work will be a good opportunity to uncover and fix issues.

Reply:
Yes, I understood the bit about `{message = x}` becoming the `grpc-message` header, that is fine as a practical solution. The bit about `body.message` becoming the _payload_ is the part that felt like an accidental side-effect. That is the part that I'd like to have changed before merging so that we don't have to keep supporting throughout 1.x for backward compatibility. I'll try to push a commit for this in this branch.